pkg/auth/dependency/sso/auth_handler_html.go

```diff
@@ -21,28 +19,98 @@ func (i *AuthHandlerHTMLProvider) HTML() (out string, err error) {
 <!DOCTYPE html>
 <html>
 <head>
-<meta name=viewport content="width=device-width, initial-scale=1">
-<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
-<script type="text/javascript" src="{{ js_sdk_cdn_url }}"></script>
 <script type="text/javascript">
-skygear.pubsub.autoPubsub = false;
-skygear.config({
-	'endPoint': '{{ api_endpoint }}',
-	'apiKey': '-'
-}).then(function(container) {
-	skygear.auth.oauthHandler();
-}, function(err) {
-	console.error(err);
-});
+function cookieJarGet(name) {
+	var jar = [];
+	var pairs = document.cookie.split("; ");
+	var i = 0;
+	for (; i < pairs.length; ++i) {
+		var parts = pairs[i].split("=");
+		var key = parts[0];
+		var value = parts.slice(1).join("=");
+		jar.push([key, value]);
+	}
+	for (var i = 0; i < jar.length; ++i) {
+		if (jar[i][0] === name) {
+			return jar[i][1];
+		}
+	}
+}
+
+function cookieJarDelete(name) {
+	document.cookie = name + "=;expires=Thu, 01 Jan 1970 00:00:01 GMT;";
+}
+
+function StringStartsWith(s, search) {
+	return s.substring(0, search.length) === search;
+}
+
+function validateCallbackURL(callbackURL, authorizedURLs) {
+	if (!callbackURL) {
+		return false;
+	}
+	if (authorizedURLs.length <= 0) {
+		return true;
+	}
+	for (var i = 0; i < authorizedURLs.length; ++i) {
+		if (StringStartsWith(callbackURL, authorizedURLs[i])) {
+			return true;
+		}
+	}
+	return false;
+}
+
+function postSSOResultMessageToWindow(windowObject, authorizedURLs) {
+	var resultStr = cookieJarGet("sso_data");
+	cookieJarDelete("sso_data");
+	var data = resultStr && JSON.parse(atob(resultStr));
+	var callbackURL = data && data.callback_url;
+	var result = data && data.result;
+	var error = null;
+	if (!result) {
+		error = 'Fail to retrieve result';
+	} else if (!callbackURL) {
+		error = 'Fail to retrieve callbackURL';
+	} else if (!validateCallbackURL(callbackURL, authorizedURLs)) {
+		error = "Unauthorized callback URL: " + callbackURL;
+	}
+
+	if (error) {
+		windowObject.postMessage({
+			type: "error",
+			error: error
+		}, "*");
+	} else {
+		windowObject.postMessage({
+			type: "result",
+			result: result
+		}, callbackURL);
```

Comment:
I think postMessage takes allowed origin, not a URL?

Reply:
Also original code.

Reply:
https://gist.github.com/louischan-oursky/bcab4ce06c30a47f775194505f7f1aa9

Tested on Chrome.